geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/metadata/FileBasedMetadata.scala

```diff
@@ -328,37 +340,65 @@ class FileBasedMetadata(
     }
   }
 
+  /**
+   * Delete a seq of paths
+   *
+   * @param paths paths to delete
+   * @param threads number of threads to use
+   */
+  private def delete(paths: Iterable[Path], threads: Int): Unit = {
+    if (threads < 2) {
+      paths.foreach(fc.delete(_, false))
+    } else {
+      val ec = ExecutionContext.fromExecutorService(new CachedThreadPool(threads))
+      try {
+        val parPaths = paths.par
+        parPaths.tasksupport = new ExecutionContextTaskSupport(ec)
+        parPaths.foreach(fc.delete(_, false))
+      } finally {
+        ec.shutdown()
+      }
+    }
+  }
+
   private class DirectoryWorker(
       es: ExecutorService,
       phaser: Phaser,
-      dir: Path,
+      listDirectory: => RemoteIterator[FileStatus],
       result: ConcurrentHashMap[String, PartitionFiles]
     ) extends Runnable {
 
     override def run(): Unit = {
       try {
-        val iter = fc.listStatus(dir)
-        while (iter.hasNext) {
+        var i = phaser.getRegisteredParties + 1
+        val iter = listDirectory
+        while (iter.hasNext && i < PhaserUtils.MaxParties) {
           val status = iter.next
           val path = status.getPath
           lazy val name = path.getName
           if (status.isDirectory) {
-            phaser.register() // register the new worker thread
-            es.submit(new DirectoryWorker(es, phaser, path, result))
+            i += 1
+            // use a tiered phaser on each directory avoid the limit of 65535 registered parties
+            es.submit(new DirectoryWorker(es, new Phaser(phaser, 1), fc.listStatus(path), result))
```

Comment:
Does this cover the situation where there is more than one directory with over 65K children?  And what happens if there are 65K+ files in the root directory?

Reply:
yeah, it will list out the first 65k and then create a new phaser to handle the next batch